src/krb5_plugin/idp/idp_kdcpreauth.c

```diff
@@ -0,0 +1,807 @@
+        return;
+    }
+
+    free(state);
+}
+
+static int
+sss_idpkdc_flags(krb5_context kctx,
+                 krb5_preauthtype pa_type)
+{
+    return PA_REPLACES_KEY;
+}
+
+static void
+sss_idpkdc_edata(krb5_context kctx,
+                 krb5_kdc_req *request,
+                 krb5_kdcpreauth_callbacks cb,
+                 krb5_kdcpreauth_rock rock,
+                 krb5_kdcpreauth_moddata moddata,
+                 krb5_preauthtype pa_type,
+                 krb5_kdcpreauth_edata_respond_fn respond,
+                 void *arg)
+{
+    struct sss_idpkdc_config *config = NULL;
+    struct sss_idpkdc_state *state;
+    char *configstr = NULL;
+    krb5_error_code ret;
+
+    state = (struct sss_idpkdc_state *)moddata;
+
+    ret = sss_idpkdc_enabled(kctx, cb, rock, &configstr);
+    if (ret != 0) {
+        goto done;
+    }
+
+    ret = sss_idpkdc_config_init(state, kctx, cb->client_name(kctx, rock),
+                                 configstr, &config);
+    if (ret != 0) {
+        goto done;
+    }
+
+    ret = sss_idpkdc_challenge_send(kctx, cb->event_context(kctx, rock), cb,
+                                    rock, respond, arg, config);
+
+done:
+    if (ret != 0) {
+        respond(arg, ret, NULL);
+    }
+
+    cb->free_string(kctx, rock, configstr);
+    sss_idpkdc_config_free(config);
+}
+
+static void
+sss_idpkdc_verify(krb5_context kctx,
+                  krb5_data *req_pkt,
+                  krb5_kdc_req *request,
+                  krb5_enc_tkt_part *enc_tkt_reply,
+                  krb5_pa_data *pa,
+                  krb5_kdcpreauth_callbacks cb,
+                  krb5_kdcpreauth_rock rock,
+                  krb5_kdcpreauth_moddata moddata,
+                  krb5_kdcpreauth_verify_respond_fn respond,
+                  void *arg)
+{
+    struct sss_idpkdc_state *state;
+    struct sss_idpkdc_config *config = NULL;
+    char *configstr = NULL;
+    krb5_error_code ret;
+    krb5_data rstate;
+
+    state = (struct sss_idpkdc_state *)moddata;
+
+    ret = sss_idpkdc_enabled(kctx, cb, rock, &configstr);
+    if (ret != 0) {
+        goto done;
+    }
+
+    ret = sss_idpkdc_config_init(state, kctx, cb->client_name(kctx, rock),
+                                 configstr, &config);
+    if (ret != 0) {
+        goto done;
+    }
+
+    ret = sss_idpkdc_get_cookie(kctx, cb, rock, &rstate);
+    if (ret != 0) {
+        goto done;
+    }
+
+    if (pa->pa_type != SSSD_IDP_OAUTH2_PADATA || pa->length != 0) {
+        ret = KRB5_PREAUTH_BAD_TYPE;
+        goto done;
+    }
+
+    /* config is freed by verify_done */
+    ret = sss_idpkdc_verify_send(kctx, cb->event_context(kctx, rock), rock,
+                                 cb, enc_tkt_reply, respond, arg,
+                                 &rstate, config);
+    if (ret != 0) {
+        goto done;
+    }
+
+    ret = 0;
+
+done:
+    if (ret != 0) {
+        respond(arg, ret, NULL, NULL, NULL);
+    }
+
+    cb->free_string(kctx, rock, configstr);
+}
```

Comment:
Coverity:
```
[Error: RESOURCE_LEAK (]()[CWE-772](https://cwe.mitre.org/data/definitions/772.html)): [[#def3]](https://cov01.lab.eng.brq.redhat.com/covscanhub/task/242632/log/added.html#def3) [important]
sssd-2.6.2/src/krb5_plugin/idp/idp_kdcpreauth.c:706: alloc_arg: "sss_idpkdc_config_init" allocates memory that is stored into "config".
sssd-2.6.2/src/krb5_plugin/idp/idp_kdcpreauth.c:738: leaked_storage: Variable "config" going out of scope leaks the storage it points to.
#  736|   
#  737|       cb->free_string(kctx, rock, configstr);
#  738|-> }
#  739|   
#  740|   static krb5_error_code
```

Reply:
If this does not silent it, it is false positive, I think.

```diff
diff --git a/src/krb5_plugin/idp/idp_kdcpreauth.c b/src/krb5_plugin/idp/idp_kdcpreauth.c
index 053887750..760fef9c5 100644
--- a/src/krb5_plugin/idp/idp_kdcpreauth.c
+++ b/src/krb5_plugin/idp/idp_kdcpreauth.c
@@ -529,6 +529,7 @@ sss_idpkdc_verify_send(krb5_context kctx,
 
     state = malloc(sizeof(struct sss_idpkdc_verify));
     if (state == NULL) {
+        sss_idpkdc_config_free(config);
         ret = ENOMEM;
         goto done;
     }

```

Reply:
Hi,

there are still the `goto done;` at lines 717 and 722 where `config` is already allocated, so I think `sss_idpkdc_config_free(config);` is needed in the `if (ret != 0)` block as well.

bye,
Sumit

Reply:
Thanks. New diff:

```diff
diff --git a/src/krb5_plugin/idp/idp_kdcpreauth.c b/src/krb5_plugin/idp/idp_kdcpreauth.c
index 9d206826f..0fd74d941 100644
--- a/src/krb5_plugin/idp/idp_kdcpreauth.c
+++ b/src/krb5_plugin/idp/idp_kdcpreauth.c
@@ -531,7 +531,6 @@ sss_idpkdc_verify_send(krb5_context kctx,
 
     state = malloc(sizeof(struct sss_idpkdc_verify));
     if (state == NULL) {
-        sss_idpkdc_config_free(config);
         ret = ENOMEM;
         goto done;
     }
@@ -565,6 +564,8 @@ sss_idpkdc_verify_send(krb5_context kctx,
 
 done:
     if (ret != 0) {
+        /* It is the caller responsibility to free config in case of error. */
+        state->config = NULL;
         sss_idpkdc_verify_free(state);
     }
 
@@ -722,7 +723,7 @@ sss_idpkdc_verify(krb5_context kctx,
         goto done;
     }
 
-    /* config is freed by verify_done */
+    /* config is freed by verify_done if ret == 0 */
     ret = sss_idpkdc_verify_send(kctx, cb->event_context(kctx, rock), rock,
                                  cb, enc_tkt_reply, respond, arg,
                                  &rstate, config);
@@ -734,6 +735,7 @@ sss_idpkdc_verify(krb5_context kctx,
 
 done:
     if (ret != 0) {
+        sss_idpkdc_config_free(config);
         respond(arg, ret, NULL, NULL, NULL);
     }
 

```